### PR TITLE
Introduce payload overflow config as a service annotation

### PR DESCRIPTION
**Pull request by arukshani**
## Automation tests
 - Unit tests 
   > Added

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes
 - Ran FindSecurityBugs plugin and verified report? yes
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes

```ballerina
@http:ServiceConfig {
    basePath: "/customOverflow",
    overflowConfig: { memoryThresholdInMB: 0.00002, tempLocation: "testBallerinaOverFlow" }
}
service<http:Service> customOverflow bind mockEP {
    @http:ResourceConfig {
        methods: ["POST"],
        path: "/"
    }
    test1(endpoint caller, http:Request request) {
        http:Response response = new;
        mime:Entity responseEntity = new;
        match request.getByteChannel() {
            error err => io:print(err.message);
            io:ByteChannel byteChannel => responseEntity.setByteChannel(byteChannel);
        }
        response.setEntity(responseEntity);
        _ = caller->respond(response);
    }
}
```

### CODE REVIEW

File: tests/ballerina-test/src/test/java/org/ballerinalang/test/services/configuration/OverflowConfigurationTest.java [hunk 1]
```diff
@@ -0,0 +1,133 @@
+package org.ballerinalang.test.services.configuration;
```
Comment (anupama-pathirage) [inline tests/ballerina-test/src/test/java/org/ballerinalang/test/services/configuration/OverflowConfigurationTest.java]: Missing license header

File: tests/ballerina-test/src/test/java/org/ballerinalang/test/services/configuration/OverflowConfigurationTest.java [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+        try {
+            String content = "ballerina";
+            HTTPTestRequest cMsg = MessageUtils.generateHTTPMessage(path, "POST", content);
+            HTTPCarbonMessage response = Services.invokeNew(serviceResult, MOCK_ENDPOINT_NAME, cMsg);
+            Assert.assertNotNull(response, "Response message not found");
+            InputStream inputStream = new HttpMessageDataStreamer(response).getInputStream();
+            Assert.assertNotNull(inputStream, "Inputstream is null");
+            ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
+            MimeUtil.writeInputToOutputStream(inputStream, outputStream);
+            Assert.assertEquals(outputStream.size(), content.getBytes().length);
```
Comment (anupama-pathirage) [inline tests/ballerina-test/src/test/java/org/ballerinalang/test/services/configuration/OverflowConfigurationTest.java]: Above code block is duplicated in multiple places in this file. Can we add a new method for that block?

File: stdlib/ballerina-mime/src/main/java/org/ballerinalang/mime/util/MimeUtil.java [hunk 3]
```diff
@@ -443,6 +460,44 @@ static boolean isNestedPartsAvailable(BStruct bodyPart) {
...
     }
 
+    /**
+     * Extract overflow configuration settings from server annotations or get the default settings.
+     *
+     * @param context   Represent ballerina context
+     * @param configAnn Server configs
+     * @return A map with overflow settings
+     */
+    public static Map<String, Object> getOverflowSettings(Context context, Annotation configAnn) {
```
Comment (chamil321) [inline stdlib/ballerina-mime/src/main/java/org/ballerinalang/mime/util/MimeUtil.java]: Shall we remove Context parameter as it has not been used anywhere within the function?

Comment (chamil321) [inline stdlib/ballerina-mime/src/main/java/org/ballerinalang/mime/util/MimeUtil.java]: Shall we remove Context parameter as it has not been used anywhere within the function?

File: stdlib/ballerina-mime/src/main/java/org/ballerinalang/mime/util/MimeUtil.java [hunk 4]
```diff
@@ -443,6 +460,44 @@ static boolean isNestedPartsAvailable(BStruct bodyPart) {
...
+        Struct overflowConfig = configAnn.getValue().getRefField(ANN_CONFIG_OVERFLOW_CONFIG).getStructValue();
+        double memoryThreshold = overflowConfig.getFloatField(ANN_CONFIG_MEMORY_THRESHOLD);
+        memoryThreshold = memoryThreshold * (ONE_MB_IN_BYTES);
+        return Double.valueOf(memoryThreshold).longValue();
+    }
+
+    private static Map<String, Object> getDefaultOverflowSettings() {
+        Map<String, Object> overflowSettings = new HashMap<>();
+        overflowSettings.put(ANN_CONFIG_MEMORY_THRESHOLD, (long) BYTE_LIMIT);
+        return overflowSettings;
```
Comment (chamil321) [inline stdlib/ballerina-mime/src/main/java/org/ballerinalang/mime/util/MimeUtil.java]: Shall we use singletonMap method to make things efficient?
return Collections.singletonMap(ANN_CONFIG_MEMORY_THRESHOLD, (long) BYTE_LIMIT);

File: tests/ballerina-test/src/test/java/org/ballerinalang/test/services/configuration/OverflowConfigurationTest.java (hunk 1 above)
Comment (arukshani) [reply tests/ballerina-test/src/test/java/org/ballerinalang/test/services/configuration/OverflowConfigurationTest.java]: Fixed it.

File: stdlib/ballerina-http/src/main/ballerina/http/annotation.bal [hunk 5]
```diff
@@ -29,6 +29,8 @@ documentation {
     F{{cors}} The cross origin resource sharing configurations for the service
     F{{versioning}} The version of the service to be used
     F{{authConfig}} Authentication configurations for securing the service
+    F{{overflowConfig}} Payload overflow configuration to control payload location(memory/temp file) based on a given
```
Comment (pubudu91) [inline stdlib/ballerina-http/src/main/ballerina/http/annotation.bal]: What does it mean if you say "payload is overflowing"? Shall we elaborate a bit on this?

File: stdlib/ballerina-http/src/main/ballerina/http/annotation.bal [hunk 6]
```diff
@@ -38,6 +40,18 @@ public type HttpServiceConfig {
     CorsConfig cors,
     Versioning versioning,
     ListenerAuthConfig? authConfig,
+    PayloadOverflowConfig overflowConfig,
+};
+
+documentation {
+    Configuration to control payload location(memory/temp file) based on a given threshold value.
```
Comment (pubudu91) [inline stdlib/ballerina-http/src/main/ballerina/http/annotation.bal]: Shall we be a bit more descriptive about this feature here? For example, what this is, why and when is it needed, how this behaves (e.g., what does Ballerina do with the saved payloads, what happens to them when they are not needed anymore)

**Comment by arukshani** [comment]
This PR is no longer valid >> https://groups.google.com/forum/#!topic/ballerina-dev/iKx6c-MJyoA
hence closing

**Pull request closed without merging**